### PR TITLE
General development branch

### PR DESCRIPTION
**Pull request by kyleam**
ATM topics are pretty dependent on each other and things are unstable,
which makes topic-based PRs difficult.  Updates will be pushed to this
general development branch so that there's a convenient place to
review changes.

### CODE REVIEW

File: pyout.py [hunk 1]
```diff
@@ -33,7 +30,7 @@ class Tabular(object):
     style : dict, optional
         Each top-level key should be a column name and the value
         should be a style dict that overrides the `default_style`
-        class attribute.
+        class attribute.  See the "Examples" section below.
```
Comment (yarikoptic) [inline pyout.py]: Examples are good. But may be add a module level constant `STYLE_ATTRIBUTES` listing all the allowed attributes (color, bold, ...)? and then include (or reference) here?

File: pyout.py [hunk 2]
```diff
@@ -43,19 +40,39 @@ class attribute.
...
+    Write another row, overriding the style:
+
+    >>> out({"name": "bar", "status": "BAD"},
+    ...     style={"status": {"color": "red", "bold": True}})
     """
 
     # TODO: Support things like auto-width, value-based coloring, etc.
     default_style = {"align": "<",
                      "width": 10,
                      "attrs": []}
```
Comment (yarikoptic) [inline pyout.py]: so is there still `attrs`?  I thought it is getting dissolved entirely from the level of API (i.e. inside attrs would be composed to be passed when dealing with Terminal backend)

File: pyout.py [hunk 3]
```diff
@@ -43,19 +40,39 @@ class attribute.
...
+
+    >>> out({"name": "foo", "status": "OK"})
+
+    Write another row, overriding the style:
+
+    >>> out({"name": "bar", "status": "BAD"},
+    ...     style={"status": {"color": "red", "bold": True}})
     """
 
     # TODO: Support things like auto-width, value-based coloring, etc.
```
Comment (yarikoptic) [inline pyout.py]: +1 ;)  for "auto-width" we might want to be able to say "minimal-width to be 10, maximal width 20, if exceeds - truncate on the right (or on the left, or in the middle, probably replaacing with ...)... or may be even make it scroll within those 20 (back and forth... I think it was done in some progress bar library that way). Needless to say that some options would not be available if output should be rendered once (e.g. when piping to the file)

Comment (yarikoptic) [reply pyout.py]: for value-based coloring -- may be it should be generalized into "value-based styling"? (e.g. making things bold when reaching some threshold) might generalize implementation etc

File: pyout.py [hunk 4]
```diff
@@ -68,31 +85,67 @@ def _build_format(self, style):
...
+            Attribute key (e.g., "color") and value (e.g., "green")
+
+        Returns
+        -------
+        str (attribute value) or None
+        """
+        if key in ["bold", "underline"]:
+            if value:
+                return key
+        if key == "color":
```
Comment (yarikoptic) [inline pyout.py]: `elif` so not to even bother to check since previous one could proceed if value is None

File: pyout.py [hunk 5]
```diff
@@ -22,9 +22,6 @@ class Tabular(object):
 
     Parameters
     ----------
-    data : iterable
-        Each element should be a dictionary where keys are the column
-        names and values are the data to write.
     columns : list of str
         Column names
     id_column : int, optional
```
Comment (yarikoptic) [inline pyout.py]: better to specify by name (or better even allowing for a combination of names) of columns. e.g. `id_columns="name"` or `id_columns=('first', 'last')`

File: pyout.py [hunk 6]
```diff
@@ -22,9 +22,6 @@ class Tabular(object):
 
     Parameters
     ----------
-    data : iterable
-        Each element should be a dictionary where keys are the column
-        names and values are the data to write.
     columns : list of str
```
Comment (yarikoptic) [inline pyout.py]: By default should just get those from the records.
BUT what would be useful is to allow for a dictionary to map possibly short /lowercased record names into full names, e.g. `{'last': "Last name", 'first': "First name"}`.
This structure could also define the order of them (if `OrderedDict` or a list)

File: pyout.py (hunk 4 above)
Comment (kyleam) [reply pyout.py]: good catch

File: pyout.py (hunk 5 above)
Comment (kyleam) [reply pyout.py]: yeah.. not sure why I went with `int` there

File: pyout.py (hunk 6 above)
Comment (kyleam) [reply pyout.py]: noted: #17

File: pyout.py (hunk 1 above)
Comment (kyleam) [reply pyout.py]: right, definitely need some sort of definition of all possible attributes.  Figured I'd wait until some of the style stuff stabilized, but I can do that now if you'd like.

File: pyout.py (hunk 2 above)
Comment (kyleam) [reply pyout.py]: there is, based on the logic in 814baa2 message, but that's probably
unnecessary complication at this point.  I'll remove `attrs` from this
level.

File: pyout.py (hunk 3 above)
Comment (kyleam) [reply pyout.py]: Noted in the issue that really needs to be split up: #1

**Comment by kyleam** [comment]
> Primarily ideas to keep in mind or implement right away ;-)

great, thanks

File: pyout.py (hunk 1 above)
Comment (kyleam) [reply pyout.py]: > module level constant STYLE_ATTRIBUTES

Thinking about this more: A list (or set) of allowed attributes would
be useful if/when we want to barf if we get an unknown attribute, but
it terms of documentation, it wouldn't help describe the acceptable
values, etc.  Seems like we should just have one place where we
document the style specification (not sure where's best.. module
docstring?) and refer to it in all other places.

Comment (yarikoptic) [reply pyout.py]: agreed!  but where that place will be? eventually it should be reflected in the docs, but also in the code.
ideally it should be "functional", i.e. actual values verified against it etc. So may be `STYLE_ATTRIBUTES` could be started as a dict (eventually become some kind of full json schema?) of type
```python
STYLE_ATTRIBUTES = dict(
 color=dict(
   description="The color",
   types=(str, unicode, ...),
 ),  
 ...
)
```
? this way you could not only use it to verify but auto-compose text listing this stuff etc

Comment (kyleam) [reply pyout.py]: noted: #18

yeah, it's certainly nice in terms of avoiding stale documentation.
Can't decide if having a spec for the spec is an overkill :)

**Comment by kyleam** [comment]
Alright, moving things back to master and topic PRs now.  Any
additional comments these changes would be still be appreciated, of
course.

**Pull request merged**